openslides/utils/consumers.py

```diff
@@ -0,0 +1,306 @@
+                },
+            )
+            await self.channel_layer.group_send(
+                "site",
+                {
+                    "type": "send_notify",
+                    "incomming": content,
+                    "senderReplyChannelName": self.channel_name,
+                    "senderUserId": self.scope['user'].id or 0,
+                },
+            )
+        else:
+            await self.send_json({'error': 'invalid message'})
+
+    async def send_notify(self, event: Dict[str, Any]) -> None:
+        """
+        Send a notify message to the user.
+        """
+        user_id = self.scope['user'].id or 0
+
+        out = []
+        for item in event['incomming']:
+            users = item.get('users')
+            reply_channels = item.get('replyChannels')
+            projectors = item.get('projectors')
+            if ((isinstance(users, list) and user_id in users)
+                    or (isinstance(reply_channels, list) and self.channel_name in reply_channels)
+                    or (users is None and reply_channels is None and projectors is None)):
+                item['senderReplyChannelName'] = event.get('senderReplyChannelName')
+                item['senderUserId'] = event.get('senderUserId')
+                item['senderProjectorId'] = event.get('senderProjectorId')
+                out.append(item)
+
+        if out:
+            await self.send_json(out)
+
+    async def send_data(self, event: Dict[str, Any]) -> None:
+        """
+        Send changed or deleted elements to the user.
+        """
+        change_id = event['change_id']
+        output = []
+        changed_elements, deleted_elements = await element_cache.get_restricted_data(self.scope['user'], change_id)
+        for collection_string, elements in changed_elements.items():
+            for element in elements:
+                output.append(format_for_autoupdate(
+                    collection_string=collection_string,
+                    id=element['id'],
+                    action='changed',
+                    data=element))
+        for element_id in deleted_elements:
+            collection_string, id = split_element_id(element_id)
+            output.append(format_for_autoupdate(
+                collection_string=collection_string,
+                id=id,
+                action='deleted'))
+        await self.send_json(output)
+
+
+class ProjectorConsumer(AsyncJsonWebsocketConsumer):
+    """
+    Websocket Consumer for the projector.
+    """
+
+    groups = ['projector']
+
+    async def connect(self) -> None:
+        """
+        Adds the websocket connection to a group specific to the projector with the given id.
+        Also sends all data that are shown on the projector.
+        """
+        user = self.scope['user']
+        projector_id = self.scope["url_route"]["kwargs"]["projector_id"]
+        await self.accept()
+
+        if not await database_sync_to_async(has_perm)(user, 'core.can_see_projector'):
```

Comment:
Yes, we should keep this consistent with the site. Also sending an error code there would be great.

Reply:
You mean the comment on line 136 thar we close the connection?

@normanjaeckel what do you think?

Reply:
Yes

Reply:
I dont know.